include/mxnet/base.h

```diff
@@ -365,7 +365,9 @@ constexpr size_t kMKLDNNAlign = 64;
 namespace std {
 template<> struct hash<mxnet::Context> {
   size_t operator()(const mxnet::Context& ctx) const {
-    return (static_cast<size_t>(ctx.dev_type) << 32) | ctx.dev_id;
+    size_t dev_type = static_cast<size_t>(ctx.dev_type);
+    size_t dev_id = static_cast<size_t>(ctx.dev_id);
+    return (dev_id << 3) ^ dev_type ^ (dev_id >> 29);
```

Comment:
I don't understand. why is this a good hash function?

Reply:
How about using dmlc::HashCombine?
like this https://github.com/apache/incubator-mxnet/blob/b59943e9ca9dd4d603ffd8b2464d431dbd1d9bdb/src/operator/nn/cudnn/cudnn_algoreg-inl.h#L160

Reply:
I find that the number of collisions is quite high, more than 3% as you can see in the test case. I'm not sure if this is a good hash function. Anyways done using your suggestion.